### PR TITLE
Update optaplanner example

### PR DESCRIPTION
**Pull request by dupliaka**
Moved to 8.24.0.Final version
Removed deprecated drools calculation
Added constraints stream solver
Updated to solver config java api

### CODE REVIEW

File: optaplanner/src/main/java/com/baeldung/optaplanner/CourseSchedule.java [hunk 1]
```diff
@@ -16,36 +16,37 @@
...
 
-    @ValueRangeProvider(id = "availableRooms")
-    @ProblemFactCollectionProperty
     public List<Integer> getRoomList() {
         return roomList;
     }
 
-    @ValueRangeProvider(id = "availablePeriods")
-    @ProblemFactCollectionProperty
+
```
Comment (yurloc) [inline optaplanner/src/main/java/com/baeldung/optaplanner/CourseSchedule.java:41]:
```suggestion
```
Please avoid consecutive blank lines.

File: optaplanner/src/main/java/com/baeldung/optaplanner/Lecture.java [hunk 2]
```diff
@@ -1,14 +1,23 @@
...
 public class Lecture {
 
+    @PlanningId
+    private Long id;
     private Integer roomNumber;
     private Integer period;
 
+    public Lecture(Long i) {
+        this.id = i;
+    }
```
Comment (yurloc) [inline optaplanner/src/main/java/com/baeldung/optaplanner/Lecture.java:17]:
```suggestion
    }

```

File: optaplanner/src/test/java/com/baeldung/optaplanner/test/CourseScheduleConstraintProviderUnitTest.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+import org.junit.jupiter.api.Test;
+import org.optaplanner.test.api.score.stream.ConstraintVerifier;
+
+public class CourseScheduleConstraintProviderUnitTest {
+
+    private final ConstraintVerifier<CourseScheduleConstraintProvider, CourseSchedule> constraintVerifier =
+            ConstraintVerifier.build(new CourseScheduleConstraintProvider(), CourseSchedule.class, Lecture.class);
+
+    @Test
+    public void testCourseScheduleConstraint(){
```
Comment (yurloc) [inline optaplanner/src/test/java/com/baeldung/optaplanner/test/CourseScheduleConstraintProviderUnitTest.java:15]:
```suggestion
    public void testCourseScheduleConstraint() {
```

File: optaplanner/src/test/java/com/baeldung/optaplanner/test/OptaPlannerUnitTest.java [hunk 4]
```diff
@@ -38,13 +53,23 @@ public void test_whenCustomJavaSolver() {
...
 
+        logger.info(ScoreManager.create(solverFactory).explainScore(solvedCourseSchedule).getSummary());
+
         Assert.assertNotNull(solvedCourseSchedule.getScore());
-        Assert.assertEquals(0, solvedCourseSchedule.getScore().getHardScore());
+        Assert.assertEquals(-4, solvedCourseSchedule.getScore().getHardScore());
+
     }
+
+
```
Comment (yurloc) [inline optaplanner/src/test/java/com/baeldung/optaplanner/test/OptaPlannerUnitTest.java:74]:
```suggestion
```
Removing an extra blank line.

File: optaplanner/src/test/java/com/baeldung/optaplanner/test/OptaPlannerUnitTest.java [hunk 5]
```diff
@@ -1,25 +1,35 @@
...
     static CourseSchedule unsolvedCourseSchedule;
 
     @BeforeAll
     public static void setUp() {
 
         unsolvedCourseSchedule = new CourseSchedule();
 
-        for(int i = 0; i < 10; i++){
-            unsolvedCourseSchedule.getLectureList().add(new Lecture());
+        for(long i = 0; i < 10; i++){
```
Comment (yurloc) [inline optaplanner/src/test/java/com/baeldung/optaplanner/test/OptaPlannerUnitTest.java:31]:
```suggestion
        for(long i = 0; i < 10; i++) {
```

**Review by yurloc — COMMENTED** [review]
A few minor improvements suggested. Otherwise looks good to me.

Of course, this shouldn't be merged without also updating the [blog post](https://www.baeldung.com/opta-planner) accordingly.

**Review by yurloc — CHANGES_REQUESTED** [review]
@dupliaka The CI fails because the `optaplanner` tutorial can no longer be built using Java 8 (because OptaPlanner 8.24 is being built with Java 11).

To fix that, go to the root `pom.xml` and move two `<module>optaplanner</module>` elements from `default-first` and `integration-lite-first` profiles to `default-jdk9-and-above` and `integration-jdk9-and-above` profiles.

See https://github.com/eugenp/tutorials#profile-based-segregation for explanation of the profiles.

**Comment by dkapil** [comment]
Thank you for your contribution @dupliaka and @yurloc. PR is merged and article is now updated.

**Comment by dupliaka** [comment]
Thank you very much! I have notice that there is no mention on ConstraintStream that is provided in test_whenConstraintStreamSolver. It is a major difference since the previous version and it would be much more consistent to mention it in the blog as well. 
You can mention in Scoring section telling simply that there is a better way to define constraints in the lambda function way that allows to cash results and calculate score faster. To do so you just need to define one class (see [CourseScheduleConstraintProvider.java](https://github.com/dupliaka/tutorials/blob/3492fb12de465c280cc2595b7c9a1d08348958d5/optaplanner/src/main/java/com/baeldung/optaplanner/CourseScheduleConstraintProvider.java))

**Comment by dkapil** [comment]
Thanks @dupliaka . We have opened an internal jira to look into this further.

**Comment by dupliaka** [comment]
> Thanks @dupliaka . We have opened an internal jira to look into this further.

Great! Thanks for collaboration feel free to reach me out if you would need some additional assistance.

**Pull request merged**